.travis.yml

```diff
@@ -3,27 +3,20 @@ dist: focal
 language: python
 python: 3.8
 cache: pip
-before_install: pip install --upgrade pip setuptools six
-install: pip install black flake8
+install:
+  - pip install --upgrade pip setuptools six
+  - pip install -r requirements_pytest.txt
```

Comment:
By convention, those test requirements file are called `requirements_dev.txt`

Reply:
That file was not needed after all.

Reply:
It's better to keep a separate file for test requirements for people who actually use it, they can install it directly. We can include all our test requirements in it, but it's up to you. 😁

Reply:
I sense we have three sets of requirements:
1. requirements.txt for our code's requirements
2. requirements like black, flake8, isort that are embedded in the pre-commit yaml file
3. the Travis CI requirements pytest and its plugins

I am reluctant to install requirements that we do not need both on contributors machines and on CI machines.

Reply:
- `pre-commit` doesn't require any installation, it is self-contained
- `test-requirements.txt` is only for the user who wants to install all the test tools used in this repository
- Travis CI won't install packages with `test-requirements.txt`, we will specify which ones to install

How does this sound?
This means we will have only two requirements file:
1. Code requirements
2. Test requirements

Reply:
I believe that we are both good with our current `requirements.txt`.

That leaves open the question, what should go in the other Test requirements file.  If I was running the tests on my machine, I would just type `pytest` because I do not need coverage and I do not need subtests (whatever that is).  So, do we create a separate Test requirements file that only contains `-r requirements.txt \n pytest`?

I would prefer a called requirements_dev.txt or requirements_test.txt because:
1. we have no files in the repo that have "`-`" in the filename
2. all the requirements files should appear next to each other in a code listing.

Reply:
`flake8`, `black`, `isort`, `codespell` fall in the `requirements_dev.txt` file as those are the tools used to develop the code as well in terms of linting and formating.

We will tell Travis to install `requirements.txt` and `pytest` and its plugins.

`pytest-subtests` are used when you're having tests for multiple iterations in a given function. By default, one function equals one test for `pytest` but if the function contains loops that iterate over many values, those will be considered as different tests. `pytest` will stop if any of those loops fail and exit immediately saying that the function failed. We want to know from all the iterations which ones are going to fail. I don't know if I can explain properly on text so you would have to do some research 😅

Reply:
This is also one of the benefits of using `pipenv` for managing your environments. It separates out your common packages and dev packages and you don't have to worry about anything but that's a discussion for another day.